src/optim/adam.rs

```diff
@@ -107,15 +107,17 @@ impl<M> GradientProvider for Adam<M> {
             let v_hat = *v * (1.0 - self.cfg.betas[1].powi(self.t)).recip();
             *g = self.cfg.lr * m_hat / (v_hat.sqrt() + self.cfg.eps)
         });
-        g_t
+        Ok(g_t)
     }
 }
 
 impl<M: CanUpdateWithGradients> Optimizer<M> for Adam<M> {
     fn update(&mut self, module: &mut M, gradients: Gradients) {
         self.t = self.t.checked_add(1).unwrap();
         self.gradients = gradients;
-        module.update(self);
+        module
+            .update(self)
+            .expect("Parameter wasn't used for gradient computation");
```

Comment:
This causes a runtime panic.

Idea: module.update() should continue with the gradient updating, and maybe the error can contain a list of unused parameters? Then Optimizer.update() can return that result and user can decide whether they want to panic or not